src/java.base/share/classes/jdk/internal/foreign/abi/BufferStack.java

```diff
@@ -0,0 +1,55 @@
+package jdk.internal.foreign.abi;
+
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemorySegment;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class BufferStack {
+    private final MemorySegment backingSegment;
+    private final ReentrantLock lock = new ReentrantLock();
```

Comment:
This could go lower level and do a CAS loop directly, but probably your solution achieves a good compromise between performance and maintainability - esp. coupled with the suggestion below of only lock when we're in a virtual thread.